pkg/controller/nodes/task/resourcemanager/redis_resourcemanager.go

```diff
@@ -0,0 +1,201 @@
+	namespacedRedisSetKey := r.getNamespacedRedisSetKey(namespace)
+	stopWatch := r.getResource(namespace).metrics.(*RedisResourceManagerMetrics).RedisSizeCheckTime.Start()
+	defer stopWatch.Stop()
+	size, err := r.client.SCard(namespacedRedisSetKey).Result()
+	if err != nil {
+		logger.Errorf(ctx, "Error getting size of Redis set in metrics poller %v", err)
+		return
+	}
+	metrics := r.getResource(namespace).metrics.(*RedisResourceManagerMetrics)
+	metrics.AllocatedTokensGauge.Set(float64(size))
+	// g, err := metrics.AllocatedTokensGauge
+	// g.Set(float64(size))
+}
+
+func (r *RedisResourceManager) startMetricsGathering(ctx context.Context) {
+	wait.Until(func() {
+		for namespace := range r.namespacedResourcesMap {
+			r.pollRedis(ctx, namespace)
+		}
+	}, 10*time.Second, ctx.Done())
+	/*
+		ticker := time.NewTicker(10 * time.Second)
+		go func() {
+			for {
+				select {
+				case <-ctx.Done():
+					return
+				case <-ticker.C:
+					for _, rrToken := range r.resourceRegistryTokens {
+						r.pollRedis(ctx, rrToken)
+					}
+				}
+			}
+		}()
+	*/
+}
+
+func NewRedisResourceManagerMetrics(scope promutils.Scope) *RedisResourceManagerMetrics {
+	return &RedisResourceManagerMetrics{
+		Scope: scope,
+		RedisSizeCheckTime: scope.MustNewStopWatch("redis:size_check_time_ms",
+			"The time it takes to measure the size of the Redis Set where all utilized resource are stored", time.Millisecond),
+
+		AllocatedTokensGauge: scope.MustNewGauge("size",
+			"The number of allocation resourceRegistryTokens currently in the Redis set"),
+	}
+}
+
+func (r *RedisResourceManager) RegisterResourceQuota(ctx context.Context, namespace resourcemanager_interface.ResourceNamespace, quota int) error {
+
+	if r.client == nil {
+		err := errors.Errorf("Redis client does not exist.")
+		return err
+	}
+	config := rmConfig.GetResourceManagerConfig()
+	if quota <= 0 || quota > config.ResourceMaxQuota {
+		err := errors.Errorf("Invalid request for resource quota (<= 0 || > %v): [%v]", config.ResourceMaxQuota, quota)
+		return err
+	}
+
+	// Checking if the namespace already exists
+	// We use linear search here because this function is only called a few times
+	if _, ok := r.namespacedResourcesMap[namespace]; ok {
+		return errors.Errorf("Resource namespace already exists [%v]", namespace)
+	}
+
+	prefixedNamespace := r.getNamespacedRedisSetKey(namespace)
+	metrics := NewRedisResourceManagerMetrics(r.MetricsScope.NewSubScope(prefixedNamespace))
+
+	newResource := &Resource{
+		quota:   quota,
+		metrics: metrics,
+	}
+
+	// Add this registration to the list
+	r.namespacedResourcesMap[namespace] = newResource
+	return nil
+}
+
+func (r *RedisResourceManager) AllocateResource(ctx context.Context, namespace resourcemanager_interface.ResourceNamespace, allocationToken string) (
+	resourcemanager_interface.AllocationStatus, error) {
+
+	namespacedRedisSetKey := r.getNamespacedRedisSetKey(namespace)
```

Comment:
TODO we should check if we can use Redis transactions to implement this as a conditional update